### ISSUE TITLE
change the default file permission when building tarballs?

### ISSUE THREAD

**Issue by johnnychen94** [JuliaPackaging/PkgServer.jl#56]
In some cases when people want to set a shared version of the Julia depot path, this could trigger a permission error.

![image](https://user-images.githubusercontent.com/8684355/85938086-6280b580-b93c-11ea-8bc4-a56135f4d525.png)

![image](https://user-images.githubusercontent.com/8684355/85938082-598fe400-b93c-11ea-9d27-edf8e270b705.png)

I'm wondering if it's possible to make it `0o755`?

https://github.com/JuliaPackaging/PkgServer.jl/blob/c5a180ed441387018d5bb539055a71e03e0d998f/bin/gen_static.jl#L176

Originally reported by @yym0924 at https://discourse.juliacn.com/t/topic/3682

**Comment by StefanKarpinski** [comment]
I'm a bit unclear on what the context is here. That `chmod` call is to ensure that a tree can be deleted, changing it to `755` should not change the ability to delete the tree in case of error.

**Comment by johnnychen94** [comment]
Yeah, I didn't read it clearly and stated wrongly.

For reproducibility, I ran these in a docker container with Julia 1.4.2:

```
JULIA_DEPOT_PATH=no_pkgserver julia -e 'using Pkg; pkg"add FFTW_jll"'
JULIA_DEPOT_PATH=with_pkgserver JULIA_PKG_SERVER="pkg.julialang.org" julia -e 'using Pkg; pkg"add FFTW_jll"'
ls -li *_pkgserver/artifacts/
```

and it outputs

```
no_pkgserver/artifacts/:
total 4
3318258 drwxr-xr-x 6 root root 4096 Jun 30 05:09 91707ee1baf86e9f971d587e5379ee3956a7c5e0

with_pkgserver/artifacts/:
total 4
3318309 drwx------ 6 root root 4096 Jun 30 05:10 91707ee1baf86e9f971d587e5379ee3956a7c5e0
```

Any ideas on how this is happening? Or if there is any fix to this on the user side so that we could still share the depot_path around.

**Comment by johnnychen94** [comment]
Yeah, I didn't read it clearly and stated wrongly.

For reproducibility, I ran these in a docker container with Julia 1.4.2:

```
JULIA_DEPOT_PATH=no_pkgserver julia -e 'using Pkg; pkg"add FFTW_jll"'
JULIA_DEPOT_PATH=with_pkgserver JULIA_PKG_SERVER="pkg.julialang.org" julia -e 'using Pkg; pkg"add FFTW_jll"'
ls -li *_pkgserver/artifacts/
```

and it outputs

```
no_pkgserver/artifacts/:
total 4
3318258 drwxr-xr-x 6 root root 4096 Jun 30 05:09 91707ee1baf86e9f971d587e5379ee3956a7c5e0

with_pkgserver/artifacts/:
total 4
3318309 drwx------ 6 root root 4096 Jun 30 05:10 91707ee1baf86e9f971d587e5379ee3956a7c5e0
```

Any ideas how this is happening?

**Comment by staticfloat** [comment]
The difference is that the original tarball contains the root directory as an entity within the tarball:

```
julia> Tar.list("./from_github.tar"; raw=true)
32-element Array{Tar.Header,1}:
 Tar.Header("./", :directory, 0o755, 0, "")
 Tar.Header("./include/", :directory, 0o755, 0, "")
 Tar.Header("./include/fftw3.f", :file, 0o444, 2447, "")
 Tar.Header("./include/fftw3.f03", :file, 0o444, 54735, "")
 Tar.Header("./include/fftw3.h", :file, 0o444, 31764, "")
 ⋮
 Tar.Header("./share/", :directory, 0o755, 0, "")
 Tar.Header("./share/licenses/", :directory, 0o755, 0, "")
 Tar.Header("./share/licenses/FFTW/", :directory, 0o755, 0, "")
 Tar.Header("./share/licenses/FFTW/COPYING", :file, 0o444, 18002, "")
 Tar.Header("./share/licenses/FFTW/COPYRIGHT", :file, 0o444, 844, "")
```

Whereas the tarball from the PkgServer has been normalized and had all directory entries removed:
```
julia> Tar.list("./from_pkgserver.tar"; raw=true)
22-element Array{Tar.Header,1}:
 Tar.Header("include/fftw3.f", :file, 0o644, 2447, "")
 Tar.Header("include/fftw3.f03", :file, 0o644, 54735, "")
 Tar.Header("include/fftw3.h", :file, 0o644, 31764, "")
 Tar.Header("include/fftw3l.f03", :file, 0o644, 27053, "")
 Tar.Header("include/fftw3q.f03", :file, 0o644, 25752, "")
 ⋮
 Tar.Header("lib/pkgconfig/fftw3.pc", :file, 0o644, 243, "")
 Tar.Header("lib/pkgconfig/fftw3f.pc", :file, 0o644, 244, "")
 Tar.Header("logs/FFTW.log", :file, 0o644, 1839180, "")
 Tar.Header("share/licenses/FFTW/COPYING", :file, 0o644, 18002, "")
 Tar.Header("share/licenses/FFTW/COPYRIGHT", :file, 0o644, 844, "")
```

Because the directory that is getting filled is done so [through `create_artifact()`](https://github.com/JuliaLang/Pkg.jl/blob/master/src/Artifacts.jl#L210), and is created through `mktempdir()`, the directory is created with permissions `0o700`.  However, even if the user has a weird `umask` set, I think all directories within an artifact should be `0o755`.  I believe `tar` and `Tar.jl` will default to using `0o777 & ~umask`, which should be something like `0o755` on most systems, but I like things to be explicit.  I suggest we alter `set_readonly()` (Which is called after [artifact installation](https://github.com/JuliaLang/Pkg.jl/blob/master/src/Artifacts.jl#L230) and [package installation](https://github.com/JuliaLang/Pkg.jl/blob/master/src/Operations.jl#L730)) to set the permissions of directories to `0o755`.  Right now, it [just sets all files to `filemode(file) & ~0o222`](https://github.com/JuliaLang/Pkg.jl/blob/master/src/utils.jl#L7), but I would argue we should be more strict with directory permissions and just unconditionally set them all to `0o755`.  I can't think of a reason a package you've installed through `add` would genuinely want different permissions (I can't think of a reason for files either, but I'm willing to be more cautious there).

**Comment by StefanKarpinski** [comment]
> I can't think of a reason a package you've installed through `add` would genuinely want different permissions

Here's one: you may have installed private code on a shared system and not want other people to be able to read it.

**Comment by staticfloat** [comment]
That’s true; but I don’t think we want that to happen at Pkg.add() time due to permissions in the tarball, instead we’d just want that to be able to be set manually, after the add is finished.

**Comment by StefanKarpinski** [comment]
Sure. Mainly I just think we should respect the umask.

**Comment by johnnychen94** [comment]
I guess this is already fixed? (https://github.com/JuliaLang/Pkg.jl/issues/2053#issuecomment-701055801)

**Comment by staticfloat** [comment]
Yes, we already fixed this on 1.6.
